### PR TITLE
Release checklist update and tweaks after 0.8.16

### PR DESCRIPTION
**Pull request by cameel**
Some tweaks to the checklist based on our experience with this release.
- Docs get built automatically so looks like we no longer need separate steps for RTD. We should verify that with @chriseth though.
- In practice we never run full `tests.sh`. I updated that point to say only to run the buglist update script.
- Clarification regarding changelog and version commits.
- Warning on the release page and more detailed steps for blog posts.

### CODE REVIEW

File: ReleaseChecklist.md [hunk 1]
```diff
@@ -3,24 +3,27 @@
...
 ### Changelog
- - [ ] Sort the changelog entries alphabetically and correct any errors you notice.
- - [ ] Create a commit on a new branch that updates the ``Changelog`` to include a release date.
- - [ ] Run ``./scripts/tests.sh`` to update the bug list.
- - [ ] Create a pull request and wait for the tests, merge it.
+ - [ ] Sort the changelog entries alphabetically and correct any errors you notice. Commit it.
+ - [ ] Copy the changelog into the blog post.
+ - [ ] Update the changelog to include a release date.
+ - [ ] Run ``scripts/update_bugs_by_version.py`` to regenerate ``bugs_by_version.json`` from the changelog and ``bugs.json``.
+ - [ ] Commit changes, create a pull request and wait for the tests. Then merge it.
```
Comment (cameel) [inline ReleaseChecklist.md]: This part may need more clarification. I.e. currently we do a separate PR for version and a separate one for changelog. The old text said to `commit on a new branch` and I guess that means two PRs?

I think we would be fine with a single PR and that's what my new version says. Is that ok or are there any good reasons for two PRs?

File: ReleaseChecklist.md [hunk 2]
```diff
@@ -50,10 +53,6 @@
  - [ ] Run ``scripts/release_ppa.sh v$VERSION`` to create the PPA release (you need the relevant openssl key).
  - [ ] Wait for the ``~ethereum/ubuntu/ethereum-static`` PPA build to be finished and published for *all platforms*. SERIOUSLY: DO NOT PROCEED EARLIER!!! *After* the static builds are *published*, copy the static package to the ``~ethereum/ubuntu/ethereum`` PPA for the destination series ``Trusty``, ``Xenial`` and ``Bionic`` while selecting ``Copy existing binaries``.
 
-### Documentation
- - [ ] Build the new version on https://readthedocs.org/projects/solidity/ (select `latest` at the bottom of the page and click `BUILD`).
- - [ ] In the admin panel, select `Versions` in the menu and set the default version to the released one.
-
```
Comment (ekpyron) [inline ReleaseChecklist.md]: Maybe instead something like "Check if https://readthedocs.org/projects/solidity looks good, i.e. if stable points to the release docs"`?

File: ReleaseChecklist.md [hunk 3]
```diff
@@ -3,24 +3,27 @@
...
+ - [ ] Update the changelog to include a release date.
+ - [ ] Run ``scripts/update_bugs_by_version.py`` to regenerate ``bugs_by_version.json`` from the changelog and ``bugs.json``.
+ - [ ] Commit changes, create a pull request and wait for the tests. Then merge it.
 
 ### Create the Release
  - [ ] Create Github release page: https://github.com/ethereum/solidity/releases/new
- - [ ] On the release page, select the ``develop`` branch as new target and set tag to the new version (e.g. `v0.8.5`) (make sure you only `SAVE DRAFT` instead of `PUBLISH RELEASE` before the actual release)
+ - [ ] On the release page, select the ``develop`` branch as new target and set tag to the new version (e.g. `v0.8.5`)
+   (make sure you only `SAVE DRAFT` instead of `PUBLISH RELEASE` before the actual release).
+   Include a warning that the release is still in progress and binaries may not yet be available.
```
Comment (ekpyron) [inline ReleaseChecklist.md]: Maybe we should fix a wording, to keep this consistent?

File: ReleaseChecklist.md (hunk 2 above)
Comment (cameel) [reply ReleaseChecklist.md]: I did add something like this but I put it under "Post-release" since it's no longer something to be actively done but more like a sanity check.

File: ReleaseChecklist.md (hunk 3 above)
Comment (cameel) [reply ReleaseChecklist.md]: Wording of what specifically? The point as a whole is not like how I'd word it but it does not seem all that inconsistent to me.

But anyway, I cleaned this up a bit now - merged this with the previous checkbox, removed two parentheses that were next to each other and switched to double backticks. Please check if the new version is ok.

File: ReleaseChecklist.md (hunk 1 above)
Comment (ekpyron) [reply ReleaseChecklist.md]: The new sequence is not yet ideal in any case, since it tells you to copy the unreviewed sorted and corrected changelog into the blog post :-).
That'd be *a* reason for doing it in two steps, especially if any change would entail running ``./scripts/test.sh`` locally :-).

But since I don't see much reason for running that locally compared to trusting in CI on the ultimate PR... I guess there's not too much reason to do this in two separate steps. I still for some reason have the feeling that doing things in two steps is cleaner, but I can't come up with an objective reason, so I'm fine with merging it into one step :-).

File: ReleaseChecklist.md (hunk 3 above)
Comment (ekpyron) [reply ReleaseChecklist.md]: Sorry, should have been clearer on that - I meant that we may want to fix the precise wording of the "release is still in progress" warning, s.t. it's always the same.

File: ReleaseChecklist.md (hunk 1 above)
Comment (cameel) [reply ReleaseChecklist.md]: > it tells you to copy the unreviewed sorted and corrected changelog into the blog post :-).

Good point. I moved that to the last point in the section.

> That'd be a reason for doing it in two steps

Right, you have to remember to rerun `scripts/update_bugs_by_version.py` if the changelog is so broken that the script did not even detect the new version. Still, I think that's not a huge risk. I added a warning to make sure the new version is there.

> I still for some reason have the feeling that doing things in two steps is cleaner, but I can't come up with an objective reason, so I'm fine with merging it into one step :-).

For me actually it seems right to have these changes grouped in a single PR. They're all minor, mechanical tweaks that happen because of the new release.

And when doing them or reviewing someone's PR I always check the PR from the previous release to make sure nothing was forgotten. Having just one PR makes this easier.

File: ReleaseChecklist.md (hunk 3 above)
Comment (cameel) [reply ReleaseChecklist.md]: Ah, ok. I added what I remember from the warning you used but I'm not sure if I didn't forget anything there. Please check.

File: ReleaseChecklist.md [hunk 4]
```diff
@@ -2,28 +2,33 @@
...
+ - [ ] Commit changes, create a pull request and wait for the tests. Then merge it.
+ - [ ] Copy the changelog into the release blog post.
 
 ### Create the Release
- - [ ] Create Github release page: https://github.com/ethereum/solidity/releases/new
- - [ ] On the release page, select the ``develop`` branch as new target and set tag to the new version (e.g. `v0.8.5`) (make sure you only `SAVE DRAFT` instead of `PUBLISH RELEASE` before the actual release)
+ - [ ] Create a [release on github](https://github.com/ethereum/solidity/releases/new).
+       Set the target to the ``develop`` branch and the tag to the new version, e.g. ``v0.8.5``.
+       Include a warning that the release is still in progress and binaries may not yet be available.
+       Include the following warning: ``**The release is still in progress and the binaries may not yet be available from all sources.**``.
```
Comment (ekpyron) [inline ReleaseChecklist.md:32]: Yeah - I mean I also just made that up - fine with using anything else, but it's convenient not to have to think about it during the release, but instead to just be able to rely on some template here :-).

File: ReleaseChecklist.md [hunk 5]
```diff
@@ -2,28 +2,33 @@
...
- - [ ] Readthedocs account, access to the Solidity project
+ - [ ] gnupg key (has to be version 1, gpg2 won't work) for ``your-name@ethereum.org`` created and uploaded
  - [ ] Write access to https://github.com/ethereum/homebrew-ethereum
 
-### Documentation check
- - [ ] Run `make linkcheck` from within `docs/` and fix any broken links it finds. Ignore false positives caused by `href` anchors and dummy links not meant to work.
-
 ### Blog Post
- - [ ] Create a post on https://github.com/ethereum/solidity-blog and explain some of the new features or concepts.
+ - [ ] Create a post on [solidity-blog](https://github.com/ethereum/solidity-blog) and explain some of the new features or concepts.
```
Comment (franzihei) [inline ReleaseChecklist.md]:
```suggestion
 - [ ] Create a post on [solidity-blog](https://github.com/ethereum/solidity-blog) in the ``Releases`` category and explain some of the new features or concepts.
```

File: ReleaseChecklist.md [hunk 6]
```diff
@@ -2,28 +2,33 @@
...
+ - [ ] Create a post on [solidity-blog](https://github.com/ethereum/solidity-blog) for each important bug.
```
Comment (franzihei) [inline ReleaseChecklist.md]:
```suggestion
 - [ ] Create a post on [solidity-blog](https://github.com/ethereum/solidity-blog) in the ``Security Alerts`` category for each important bug.
```

File: ReleaseChecklist.md [hunk 7]
```diff
@@ -62,7 +63,10 @@
  - [ ] Create a tag using ``git tag --annotate v$VERSION`` and push it with ``git push --tags``.
 
 ### Post-release
- - [ ] Publish the blog post.
+ - [ ] Make sure the documentation for the new release has been published successfully.
+       Go to https://readthedocs.org/projects/solidity/ and verify that the new version is listed, works and is marked as default.
+ - [ ] Remove "still in progress" warning from the release page.
```
Comment (franzihei) [inline ReleaseChecklist.md]:
```suggestion
 - [ ] Remove "still in progress" warning from the release notes.
```

File: ReleaseChecklist.md [hunk 8]
```diff
@@ -62,7 +63,10 @@
...
+ - [ ] Publish the blog posts.
  - [ ] Create a commit to increase the version number on ``develop`` in ``CMakeLists.txt`` and add a new skeleton changelog entry.
  - [ ] Announce on Twitter, including links to the release and the blog post.
```
Comment (franzihei) [inline ReleaseChecklist.md:81]:
```suggestion
 - [ ] Announce on Twitter, including links to the release and the blog post. Use ``#xp`` at the end of the tweet to automatically cross post the announcement to Fosstodon. 
```

File: ReleaseChecklist.md [hunk 9]
```diff
@@ -62,7 +63,10 @@
...
  - [ ] Share announcement on Reddit and Solidity forum.
```
Comment (franzihei) [inline ReleaseChecklist.md]:
```suggestion
 - [ ] Share announcement on Reddit (r/ethdev) and Solidity forum at forum.soliditylang.org in the "Announcements" category.
```

**Review by franzihei — COMMENTED** [review]
Added some details re. communications

File: ReleaseChecklist.md (hunk 9 above)
Comment (cameel) [reply ReleaseChecklist.md]: What about `/r/ethereum`?

Comment (franzihei) [reply ReleaseChecklist.md]: yep... or crossposting. I think r/ethereum has grown quite generic, while r/ethdev is more technical

File: ReleaseChecklist.md (hunk 1 above)
Comment (ekpyron) [reply ReleaseChecklist.md]: Yeah, as I said, I can't come up with an objective reason not to, so I'm fine with making it one PR :-).

File: ReleaseChecklist.md (hunk 4 above)
Comment (cameel) [reply ReleaseChecklist.md:32]: Oh, just noticed that I left the old line in :) Fixed.

Comment (cameel) [reply ReleaseChecklist.md:32]: I considered actually adding a template of the release notes but decided to leave it out since this is a checklist and should be relatively short. It's probably just as convenient to copy and adjust the one from the last release anyway. It's only the warning that you have to write from scratch every time.

File: ReleaseChecklist.md (hunk 6 above)
Comment (cameel) [reply ReleaseChecklist.md]: Done.

File: ReleaseChecklist.md (hunk 7 above)
Comment (cameel) [reply ReleaseChecklist.md]: Done here and in a few other places.

File: ReleaseChecklist.md (hunk 8 above)
Comment (cameel) [reply ReleaseChecklist.md:81]: Done.

File: ReleaseChecklist.md (hunk 9 above)
Comment (cameel) [reply ReleaseChecklist.md]: ok, I added a note about cross-posting, I can still remove it if you think `/r/ethdev` is enough.

I also split forum and reddit into separate points.

File: ReleaseChecklist.md [hunk 10]
```diff
@@ -2,28 +2,32 @@
 
 ### Requirements
  - [ ] Lauchpad (Ubuntu One) account
- - [ ] gnupg key (has to be version 1, gpg2 won't work) for `your-name@ethereum.org` created and uploaded
- - [ ] Readthedocs account, access to the Solidity project
+ - [ ] gnupg key (has to be version 1, gpg2 won't work) for ``your-name@ethereum.org`` created and uploaded
```
Comment (ekpyron) [inline ReleaseChecklist.md]: Not entire clear what to do with that key... i.e. we could merge this with the previous point (Are gpg keys required for anything *but* launchpad? Nothing comes to my mind.)

File: ReleaseChecklist.md [hunk 11]
```diff
@@ -2,28 +2,32 @@
...
  - [ ] Write access to https://github.com/ethereum/homebrew-ethereum
 
-### Documentation check
- - [ ] Run `make linkcheck` from within `docs/` and fix any broken links it finds. Ignore false positives caused by `href` anchors and dummy links not meant to work.
-
 ### Blog Post
- - [ ] Create a post on https://github.com/ethereum/solidity-blog and explain some of the new features or concepts.
+ - [ ] Create a post on [solidity-blog](https://github.com/ethereum/solidity-blog) in the ``Releases`` category and explain some of the new features or concepts.
+ - [ ] Create a post on [solidity-blog](https://github.com/ethereum/solidity-blog) in the ``Security Alerts`` category for each important bug.
```
Comment (ekpyron) [inline ReleaseChecklist.md]: A blog post PR actually... and "for each important bug" is rather pessimistic :-D... ``in case of important bug(s)`` rather?

File: ReleaseChecklist.md [hunk 12]
```diff
@@ -2,28 +2,32 @@
...
 
 ### Create the Release
- - [ ] Create Github release page: https://github.com/ethereum/solidity/releases/new
- - [ ] On the release page, select the ``develop`` branch as new target and set tag to the new version (e.g. `v0.8.5`) (make sure you only `SAVE DRAFT` instead of `PUBLISH RELEASE` before the actual release)
- - [ ] Thank voluntary contributors in the Github release page (use ``git shortlog -s -n -e v0.5.3..origin/develop``).
+ - [ ] Create a [release on github](https://github.com/ethereum/solidity/releases/new).
+       Set the target to the ``develop`` branch and the tag to the new version, e.g. ``v0.8.5``.
+       Include the following warning: ``**The release is still in progress and the binaries may not yet be available from all sources.**``.
+       Don't publish it yet - click the ``Save draft`` button instead.
+ - [ ] Thank voluntary contributors in the Github release notes (use ``git shortlog -s -n -e v0.5.3..origin/develop``).
```
Comment (ekpyron) [inline ReleaseChecklist.md]:
```suggestion
 - [ ] Thank voluntary contributors in the Github release notes (use ``git shortlog -s -e v0.5.3..origin/develop``).
```

In practice we don't sort by number of commits...

**Review by ekpyron — DISMISSED** [review]
Some nitpicking comments :-). But we can also just merge.

File: ReleaseChecklist.md (hunk 10 above)
Comment (cameel) [reply ReleaseChecklist.md]: Good point. I think that's the only thing. Unless dockerhub also requires it (I don't have access to that) but I doubt it.

While at it I also added other requirements: all repos, npm, dockerhub, twitter, reddit, etc. Check it out.

File: ReleaseChecklist.md (hunk 11 above)
Comment (cameel) [reply ReleaseChecklist.md]: Haha, it's technically correct but I guess we can make it less scary :) Changed.

File: ReleaseChecklist.md (hunk 12 above)
Comment (cameel) [reply ReleaseChecklist.md]: Done. Also changed to long options so that it's easier to see what the command is supposed to do.

File: ReleaseChecklist.md (hunk 10 above)
Comment (Marenz) [reply ReleaseChecklist.md]: I mean, if enough ppl have keys, we could start signing our commits?

Comment (Marenz) [reply ReleaseChecklist.md]: (just thinking about other cases where they are useful)

File: ReleaseChecklist.md [hunk 13]
```diff
@@ -1,29 +1,39 @@
...
- - [ ] gnupg key (has to be version 1, gpg2 won't work) for `your-name@ethereum.org` created and uploaded
- - [ ] Readthedocs account, access to the Solidity project
- - [ ] Write access to https://github.com/ethereum/homebrew-ethereum
-
-### Documentation check
- - [ ] Run `make linkcheck` from within `docs/` and fix any broken links it finds. Ignore false positives caused by `href` anchors and dummy links not meant to work.
+ - [ ] Github account with access to [solidity](https://github.com/ethereum/solidity), [solc-js](https://github.com/ethereum/solc-js),
+       [solc-bin](https://github.com/ethereum/solc-bin), [homebrew-ethereum](https://github.com/ethereum/homebrew-ethereum),
+       [solidity-blog](https://github.com/ethereum/solidity-blog) and [solidity-portal](https://github.com/ethereum/solidity-portal) repositories.
+ - [ ] DockerHub account with push rights to the [``solc`` image](https://hub.docker.com/r/ethereum/solc).
```
Comment (cameel) [inline ReleaseChecklist.md:7]: I don't have dockerhub account so not sure how that works in details. Do you need access to a specific image or do you just need to be a member of the [ethereum](https://hub.docker.com/u/ethereum) organization?

File: ReleaseChecklist.md (hunk 10 above)
Comment (cameel) [reply ReleaseChecklist.md]: Probably. I wonder how that works with rebases though. Doesn't this invalidate signatures?

File: ReleaseChecklist.md (hunk 13 above)
Comment (ekpyron) [reply ReleaseChecklist.md:7]: I think being part of the ``ethereum`` organization is sufficient. But ultimately what you happens is that you'll push the ``solc`` image, so I'd say the current phrasing is fine.

**Comment by ekpyron** [comment]
I just merged https://github.com/ethereum/solidity/pull/13370, so would be nice to tweak the checklist for that here as well.

File: ReleaseChecklist.md (hunk 10 above)
Comment (ekpyron) [reply ReleaseChecklist.md]: I think the sentiment in the past was that signed commits aren't really all that useful in general :-).

File: ReleaseChecklist.md [hunk 14]
```diff
@@ -42,17 +52,18 @@
...
 
 ### PPA
- - [ ] Make sure the ``ethereum/cpp-build-deps`` PPA repository contains libz3-static-dev builds for all current versions of ubuntu. If not run ``scripts/deps-ppa/static-z3.sh`` (after changing email address and key id and adding the missing ubuntu version) and wait for the builds to succeed before continuing.
- - [ ] Change ``scripts/release_ppa.sh`` to match your key's email and key id; double-check that ``DISTRIBUTIONS`` contains the most recent versions.
+ - [ ] Create ``.release_ppa_auth`` at the root of your local Solidity checkout and set ``LAUNCHPAD_EMAIL`` and ``LAUNCHPAD_KEYID`` to your key's email and key id.
+ - [ ] Double-check that the ``DISTRIBUTIONS`` list in ``scripts/release_ppa.sh`` and ``scripts/deps-ppa/static-z3.sh`` contains the most recent versions of Ubuntu.
+ - [ ] Make sure the [``~ethereum/cpp-build-deps`` PPA repository](https://launchpad.net/~ethereum/+archive/ubuntu/cpp-build-deps) contains ``libz3-static-dev builds`` for all current versions of Ubuntu.
+       If not, run ``scripts/deps-ppa/static-z3.sh`` (after changing email address and key id) and wait for the builds to succeed before continuing.
```
Comment (cameel) [inline ReleaseChecklist.md:61]: @Marenz I just realized that #13370 did not cover `scripts/deps-ppa/static-z3.sh` so it still has a hard-coded key and e-mail. I think we should have them both working the same way.

File: ReleaseChecklist.md [hunk 15]
```diff
@@ -42,17 +52,18 @@
...
  - [ ] Run ``scripts/release_ppa.sh v$VERSION`` to create the PPA release (you need the relevant openssl key).
```
Comment (cameel) [inline ReleaseChecklist.md]: @Marenz What key is it referencing? Does it mean GPG key or did you actually need some OpenSSL key here?

File: ReleaseChecklist.md [hunk 16]
```diff
@@ -42,17 +52,18 @@
...
- - [ ] Wait for the ``~ethereum/ubuntu/ethereum-static`` PPA build to be finished and published for *all platforms*. SERIOUSLY: DO NOT PROCEED EARLIER!!! *After* the static builds are *published*, copy the static package to the ``~ethereum/ubuntu/ethereum`` PPA for the destination series ``Trusty``, ``Xenial`` and ``Bionic`` while selecting ``Copy existing binaries``.
-
-### Documentation
- - [ ] Build the new version on https://readthedocs.org/projects/solidity/ (select `latest` at the bottom of the page and click `BUILD`).
- - [ ] In the admin panel, select `Versions` in the menu and set the default version to the released one.
+ - [ ] Wait for the [``~ethereum/ethereum-static`` PPA](https://launchpad.net/~ethereum/+archive/ubuntu/ethereum-static) build to be finished and published for *all platforms*.
+       **SERIOUSLY: DO NOT PROCEED EARLIER!!!**
+       *After* the static builds are *published*, copy the static package to the [``~ethereum/ethereum`` PPA](https://launchpad.net/~ethereum/+archive/ubuntu/ethereum)
+       for the destination series ``Trusty``, ``Xenial`` and ``Bionic`` while selecting ``Copy existing binaries``.
```
Comment (cameel) [inline ReleaseChecklist.md:66]: @Marenz Since you did this last time - is this list up to date?

File: ReleaseChecklist.md (hunk 15 above)
Comment (Marenz) [reply ReleaseChecklist.md]: Yeah, I am not sure what ssl keys are meant here, as far as I know I only needed the gpg keys registered with launchpad

File: ReleaseChecklist.md (hunk 16 above)
Comment (Marenz) [reply ReleaseChecklist.md:66]: I think the list is good. Those are the three oldest releases that are still supported one way or another and non of them appear in the script file, so they all need to be copied from the static version as described

File: ReleaseChecklist.md (hunk 15 above)
Comment (cameel) [reply ReleaseChecklist.md]: OK, I'm removing this bit in that case.

**Review by matheusaaguiar — DISMISSED** [review]

File: ReleaseChecklist.md [hunk 17]
```diff
@@ -62,9 +73,14 @@
  - [ ] Create a tag using ``git tag --annotate v$VERSION`` and push it with ``git push --tags``.
 
 ### Post-release
- - [ ] Publish the blog post.
+ - [ ] Make sure the documentation for the new release has been published successfully.
+       Go to https://readthedocs.org/projects/solidity/ and verify that the new version is listed, works and is marked as default.
```
Comment (matheusaaguiar) [inline ReleaseChecklist.md]:
```suggestion
       Go to [readthedocs/solidity](https://readthedocs.org/projects/solidity/) and verify that the new version is listed, works and is marked as default.
```

File: ReleaseChecklist.md [hunk 18]
```diff
@@ -42,17 +52,18 @@
  - [ ] Update the version and the hash (``sha256sum solidity_$VERSION.tar.gz``) in https://github.com/ethereum/homebrew-ethereum/blob/master/solidity.rb
```
Comment (matheusaaguiar) [inline ReleaseChecklist.md]:
```suggestion
 - [ ] Update the version and the hash (``sha256sum solidity_$VERSION.tar.gz``) in [homebrew-ethereum formula](https://github.com/ethereum/homebrew-ethereum/blob/master/solidity.rb)
```

Comment (matheusaaguiar) [reply ReleaseChecklist.md]: Would also suggest a similar change to the line before.

**Review by matheusaaguiar — COMMENTED** [review]
I noticed that there are 3 "raw" urls in the text while others were wrapped in hyperlinks. Probably we want to do this with these remaining ones.

File: ReleaseChecklist.md (hunk 17 above)
Comment (cameel) [reply ReleaseChecklist.md]: Done. Just used a slightly different text for the link.

File: ReleaseChecklist.md (hunk 18 above)
Comment (cameel) [reply ReleaseChecklist.md]: Done.

**Review by matheusaaguiar — APPROVED** [review]

**Review by ekpyron — APPROVED** [review]
